File: examples/python/benchmark.py
#!/usr/bin/env python3
"""JSON throughput through the Galley Python bindings: no AST, no procedures,
no error recovery. Parses languages/json/samples/code-01.json 50,000 times
on one session and reports bytes/s."""

from __future__ import annotations

import os
import sys
import time
from pathlib import Path

LOGICAL_INPUT = "languages/json/samples/code-01.json"
DEFAULT_ITERATIONS = 50_000


def resolve_input(explicit: str | None) -> Path:
    if explicit:
        return Path(explicit)
    checkout = os.environ.get("GALLEY_CHECKOUT")
    if checkout:
        candidate = Path(checkout) / LOGICAL_INPUT
        if candidate.is_file():
            return candidate
    candidate = Path(__file__).resolve().parent.parent.parent / LOGICAL_INPUT
    if candidate.is_file():
        return candidate
    raise SystemExit(f"missing {LOGICAL_INPUT}")


def main() -> int:
    arguments = sys.argv[1:]
    iterations = DEFAULT_ITERATIONS
    explicit: str | None = None
    if len(arguments) > 0:
        explicit = arguments[0]
    if len(arguments) > 1:
        iterations = int(arguments[1])
        if iterations < 1:
            print("iterations must be >= 1", file=sys.stderr)
            return 1

    language_dir = Path(__file__).resolve().parent / "benchmark"
    sys.path.insert(0, str(language_dir))
    import galley

    path = resolve_input(explicit)
    try:
        data = path.read_bytes()
    except OSError:
        print(f"failed to read {LOGICAL_INPUT}", file=sys.stderr)
        return 1

    try:
        session = galley.Session()
    except galley.Error:
        print("failed to create a parser session", file=sys.stderr)
        return 1

    with session:
        try:
            parsed = session.parse(data)
        except galley.Error as error:
            print(f"warmup parse failed: {error} ({error.code})", file=sys.stderr)
            return 1
        if parsed != len(data):
            print(
                f"warmup parse failed: parsed {parsed} of {len(data)} bytes",
                file=sys.stderr,
            )
            return 1

        start = time.perf_counter_ns()
        for index in range(iterations):
            try:
                parsed = session.parse(data)
            except galley.Error as error:
                print(
                    f"parse failed at iteration {index}: {error} ({error.code})",
                    file=sys.stderr,
                )
                return 1
            if parsed != len(data):
                print(
                    f"parse failed at iteration {index}: parsed {parsed} of {len(data)} bytes",
                    file=sys.stderr,
                )
                return 1
        elapsed = time.perf_counter_ns() - start

    total = len(data) * iterations
    bps = 0 if elapsed == 0 else total * 1_000_000_000 // elapsed
    print(f"input: {LOGICAL_INPUT}")
    print(f"bytes: {len(data):,}")
    print(f"iterations: {iterations:,}")
    print(f"parsed_bytes: {total:,}")
    print(f"duration_ns: {elapsed:,}")
    print(f"bytes_per_second: {bps:,}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/js/benchmark.ts
#!/usr/bin/env node
/**
 * JSON throughput through the Galley TypeScript bindings: no AST, no
 * procedures, no error recovery. Parses languages/json/samples/code-02.json
 * 10 times on one session and reports bytes/s.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import process from "node:process";
import { galley } from "@sanbus/galley";
import type { Session } from "@sanbus/galley";
import * as json from "./json/index.mjs";

const LOGICAL_INPUT = "languages/json/samples/code-02.json";
const DEFAULT_ITERATIONS = 10;

function resolveInput(explicit: string | undefined): string {
  if (explicit) return explicit;
  const checkout = process.env.GALLEY_CHECKOUT;
  if (checkout) {
    const candidate = path.join(checkout, LOGICAL_INPUT);
    if (fs.existsSync(candidate)) return candidate;
    console.error(`GALLEY_CHECKOUT=${checkout} has no ${LOGICAL_INPUT}`);
    process.exit(1);
  }
  console.error(`pass the sample file or set GALLEY_CHECKOUT at a Galley checkout (needs ${LOGICAL_INPUT})`);
  process.exit(1);
}

// GALLEY_WASM=1 (or a path to a `.wasm` module file) benchmarks the
// WebAssembly backend instead of native, mirroring demo.ts.
function jsonWasmBytes(): Uint8Array | "dir" | null {
  const selected = process.env.GALLEY_WASM;
  if (!selected) return null;
  process.env.GALLEY_QUIET = "1";
  if (selected === "1") return "dir";
  return new Uint8Array(fs.readFileSync(selected));
}

async function main(): Promise<number> {
  const arguments_ = process.argv.slice(2);
  let iterations = DEFAULT_ITERATIONS;
  const explicit = arguments_[0];
  if (arguments_.length > 1) {
    iterations = Number.parseInt(arguments_[1], 10);
    if (!Number.isInteger(iterations) || iterations < 1) {
      console.error("iterations must be >= 1");
      return 1;
    }
  }

  const filePath = resolveInput(explicit);
  let data: Buffer;
  try {
    data = fs.readFileSync(filePath);
  } catch {
    console.error(`failed to read ${LOGICAL_INPUT}`);
    return 1;
  }
  const length = data.length;

  let session: Session;
  try {
    await json.initialize();
    const wasm = jsonWasmBytes();
    if (wasm !== null && wasm !== "dir") {
      const language = await galley.loadBytes(wasm);
      session = await language.openSession();
    } else {
      session =
        wasm === null
          ? await json.openSession()
          : await json.openSession({ backend: "wasm" });
    }
  } catch {
    console.error("failed to create a parser session");
    return 1;
  }

  try {
    let parsed: number;
    try {
      parsed = session.parse(data);
    } catch (err: unknown) {
      console.error(`warmup parse failed: ${err}`);
      return 1;
    }
    if (parsed !== length) {
      console.error(`warmup parse failed: parsed ${parsed} of ${length} bytes`);
      return 1;
    }

    const start = process.hrtime.bigint();
    let index = 0;
    try {
      for (; index < iterations; index++) {
        parsed = session.parse(data);
        if (parsed !== length) break;
      }
    } catch (err: unknown) {
      console.error(`parse failed at iteration ${index}: ${err}`);
      return 1;
    }
    const elapsed = process.hrtime.bigint() - start;
    if (parsed !== length) {
      console.error(`parse failed at iteration ${index}: parsed ${parsed} of ${length} bytes`);
      return 1;
    }
    const total = BigInt(length) * BigInt(iterations);
    const bps = elapsed === 0n ? 0n : (total * 1_000_000_000n) / elapsed;

    console.log(`input: ${LOGICAL_INPUT}`);
    console.log(`bytes: ${withThousands(length)}`);
    console.log(`iterations: ${withThousands(iterations)}`);
    console.log(`parsed_bytes: ${withThousands(total)}`);
    console.log(`duration_ns: ${withThousands(elapsed)}`);
    console.log(`bytes_per_second: ${withThousands(bps)}`);
    return 0;
  } finally {
    session.close();
  }
}

function withThousands(n: number | bigint): string {
  const digits = n.toString();
  let out = "";
  for (let i = 0; i < digits.length; i++) {
    if (i > 0 && (digits.length - i) % 3 === 0) out += ",";
    out += digits[i];
  }
  return out;
}

main().then((code) => process.exit(code));
